src/transform/src/predicate_pushdown.rs

```diff
@@ -84,357 +82,411 @@ impl PredicatePushdown {
+                x.visit1_mut(|e| self.dataflow_transform(e, get_predicates));
+                // Prevent local predicate lists from escaping.
+                if let RelationExpr::Let { id, .. } = x {
+                    get_predicates.remove(&Id::Local(*id));
+                }
+            }
+        }
     }
 
-    /// Single node predicate pushdown
+    /// Predicate pushdown
+    ///
+    /// This method looks for opportunities to push predicates toward
+    /// sources of data. Primarily, this is the `Filter` expression,
+    /// and moving its predicates through the operators it contains.
+    ///
+    /// In addition, the method accumulates the intersection of predicates
+    /// applied to each `Get` expression, so that the predicate can
+    /// then be pushed through to a `Let` binding, or to the external
+    /// source of the data if the `Get` binds to another view.
     fn action(
         &self,
         relation: &mut RelationExpr,
         get_predicates: &mut HashMap<Id, HashSet<ScalarExpr>>,
     ) {
-        if let RelationExpr::Filter { input, predicates } = relation {
-            match &mut **input {
-                RelationExpr::Let { id, value, body } => {
-                    // Push all predicates to the body.
-                    **body = body
-                        .take_dangerous()
-                        .filter(std::mem::replace(predicates, Vec::new()));
+        // In the case of Filter or Get we have specific work to do;
+        // otherwise we should recursively descend.
+        match relation {
+            RelationExpr::Filter { input, predicates } => {
+                // Depending on the type of `input` we have different
+                // logic to apply to consider pushing `predicates` down.
+                match &mut **input {
+                    RelationExpr::Let { id, value, body } => {
+                        // Push all predicates to the body.
+                        **body = body
+                            .take_dangerous()
+                            .filter(std::mem::replace(predicates, Vec::new()));
 
-                    // A `Let` binding wants to push all predicates down in
-                    // `body`, and collect the intersection of those pushed
-                    // at `Get` statements. The intersection can be pushed
-                    // down to `value`.
-                    self.action(body, get_predicates);
-                    if let Some(list) = get_predicates.remove(&Id::Local(*id)) {
-                        // `list` contains the intersection of predicates.
-                        body.visit_mut(&mut |e| {
-                            if let RelationExpr::Filter { input, predicates } = e {
-                                if let RelationExpr::Get { id: id2, .. } = **input {
-                                    if id2 == Id::Local(*id) {
-                                        predicates.retain(|p| !list.contains(p));
-                                    }
```

Comment:
This block is being removed due to redundancy?

Reply:
I think this block was the before because we might not otherwise have determine which predicates we should use. But I think this logic was wrong also, for the same reasons. I'm not entirely sure what led the code to be in this state, but I think the changes are a move towards more correct!